### PR TITLE
MAT-6917:  Frontend for Retrieving codes and versions, add ValueSets tab behind a featureFlag

### PR DESCRIPTION
**Pull request by mcmcphillips**
## MADiE PR

Jira Ticket: [MAT-6917](https://jira.cms.gov/browse/MAT-6917)
(Optional) Related Tickets:
Jira Ticket: [MAT-6389](https://jira.cms.gov/browse/MAT-6389)

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included
* [x] No extemporaneous files are included (i.e Complied files or testing results)
* [x] This PR is in to the **correct branch**.
* [x] All Documentation as needed for this PR is Complete (or noted in a TODO or other Ticket)
* [x] Any breaking changes or failing automation are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package)
* [ ] All CDN/Web dependancies are hosted internally (i.e MADiE-Root Repo)

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
*  The tests appropriately test the new code, including edge cases
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
+    "^.+\\.js?$": require.resolve("babel-jest"),
   },
+  transformIgnorePatterns: ["node_modules/(?!formik)/"],
   moduleNameMapper: {
-    "^axios$": require.resolve("axios"),
     "\\.(css)$": "identity-obj-proxy",
     "single-spa-react/parcel": "single-spa-react/lib/cjs/parcel.cjs",
+    "^react($|/.+)": "<rootDir>/node_modules/react$1", // makes sure all React imports are running off of the one in this package.
     "\\.(css|scss)$": "identity-obj-proxy",
+    "@madie/madie-components":
```
Comment (RohitKandimalla) [inline jest.config.js]: We don't need this anymore I believe. This project is archieved.

File: src/api/useTerminologyServiceApi.ts [hunk 2]
```diff
@@ -110,6 +120,41 @@ export class TerminologyServiceApi {
       );
     }
   }
+
+  // uncomment if you need a way to request all the codeSystems from terminology.
+  // async updateCodeSystems(): Promise<any> {
```
Comment (RohitKandimalla) [inline src/api/useTerminologyServiceApi.ts]: I don't think we will be needing this from our madie-editor

File: src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx [hunk 3]
```diff
@@ -6,8 +6,26 @@ import {
   waitFor,
 } from "@testing-library/react";
 import CqlEditorWithTerminology from "./CqlEditorWithTerminology";
+import axios from "axios";
 import React from "react";
+import { useServiceConfig } from "../api/useServiceConfig";
```
Comment (RohitKandimalla) [inline src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]: We don't need this import as we are mock the entire module ?

File: src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx [hunk 4]
```diff
@@ -6,8 +6,26 @@ import {
   waitFor,
 } from "@testing-library/react";
 import CqlEditorWithTerminology from "./CqlEditorWithTerminology";
+import axios from "axios";
 import React from "react";
```
Comment (RohitKandimalla) [inline src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]:
```suggestion
import * as React from "react";
```

File: src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx [hunk 5]
```diff
@@ -6,8 +6,26 @@ import {
...
+import { useServiceConfig } from "../api/useServiceConfig";
 
+jest.mock("axios");
+const mockedAxios = axios as jest.Mocked<typeof axios>;
+
+const mockConfig: ServiceConfig = {
```
Comment (RohitKandimalla) [inline src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx:28]: We may want to import ServiceConfig interface ?

File: src/terminologySectionPanel/TerminologySectionPanel.tsx [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import React, { useState } from "react";
+import React, { useState, useRef, useEffect } from "react";
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/TerminologySectionPanel.tsx]: No other changes in this file, Doesn't look like need these imports.

File: src/terminologySectionPanel/codesSection/CodesSection.test.tsx [hunk 7]
```diff
@@ -8,7 +8,41 @@ import {
 } from "@testing-library/react";
 import * as React from "react";
 import CodesSection from "./CodesSection";
+import axios from "axios";
+import { useServiceConfig } from "../../api/useServiceConfig";
+import { useCodeSystems } from "./useCodeSystems";
+import useTerminologyServiceApi from "../../api/useTerminologyServiceApi";
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/CodesSection.test.tsx]: Few imports we don't need, and some unused mocks like axios

File: src/terminologySectionPanel/codesSection/CodesSection.test.tsx [hunk 8]
```diff
@@ -8,7 +8,41 @@ import {
...
+];
+
+const mockUseCodeSystems = useCodeSystems as jest.MockedFunction<
+  typeof useCodeSystems
+>;
+mockUseCodeSystems.mockReturnValue({
+  codeSystems: mockCodeSystems,
+});
 const renderEditor = () => {
   return render(<CodesSection />);
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/CodesSection.test.tsx]: Looks like CodesSection component is expecting a canEdit prop.

File: src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSection.tsx [hunk 9]
```diff
@@ -11,51 +11,71 @@ import InputAdornment from "@material-ui/core/InputAdornment";
...
+  value: string;
+  label: string;
+}
+
 export default function CodeSection({
   handleFormSubmit,
+  allCodeSystems,
   canEdit,
-}: CodeSectionProps) {
+}) {
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSection.tsx]:
```suggestion
}: CodeSectionProps) {
```

Comment (RohitKandimalla) [reply src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSection.tsx]: May also want to define the type for allCodeSystems. CodeSystem[] probably ?

File: src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSection.tsx [hunk 10]
```diff
@@ -11,51 +11,71 @@ import InputAdornment from "@material-ui/core/InputAdornment";
...
     },
     validationSchema: CodeSubSectionSchemaValidator,
     onSubmit: (values) => {
       handleFormSubmit(values);
     },
   });
-
+  const [availableVersions, setAvailableVersions] = useState([]);
+  useEffect(() => {
+    if (formik.values.title) {
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSection.tsx:64]: Do we have to do _.isEmpty() or something ? unless I guess its fine in JS

File: src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx [hunk 11]
```diff
@@ -2,10 +2,16 @@ import React from "react";
...
+jest.mock("axios");
 
 describe("CodeSub Section component", () => {
   it("should display Codes(s) and Results sections when navigated to code tab", async () => {
-    const { getByTestId } = render(<CodeSubSection canEdit={false} />);
+    const { getByTestId } = render(
+      <CodeSubSection canEdit={false} allCodeSystems={mockedCodeSystems} />
+    );
 
     const codeSubTabHeading = await getByTestId(
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx]: We don't need await for getByTestId

File: src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx [hunk 12]
```diff
@@ -40,6 +48,11 @@ describe("CodeSub Section component", () => {
       "code-system-version-selector-input"
     );
     expect(codeSystemVersionSelectInput).toBeInTheDocument();
+    expect(codeSystemVersionSelectInput.value).toBe("2.0");
+    fireEvent.change(codeSystemSelectInput, {
+      target: { value: "1.0" },
+    });
+    expect(codeSystemVersionSelectInput.value).toBe("2.0");
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx:60]: I found that if we have use attribute on HTML elements, our linter is happy when we declare the Element, Something like 

const codeSystemVersionSelectInput = getByTestId(
      "code-system-version-selector-input"
    ) as HTMLInputElement ;

File: src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx [hunk 13]
```diff
@@ -55,7 +68,28 @@ describe("CodeSub Section component", () => {
...
+      <CodeSubSection allCodeSystems={mockedCodeSystems} canEdit={true} />
+    );
+
+    const clearButton = getByTestId("clear-codes-btn");
+    expect(clearButton).toBeDisabled();
+
+    const codeText = getByTestId("code-text");
+    expect(codeText).toBeEnabled();
+    expect(codeText).toBeInTheDocument();
+    const codeTextInput = getByTestId("code-text-input");
```
Comment (RohitKandimalla) [inline src/terminologySectionPanel/codesSection/codesSubSection/codeSubSection/CodeSubSection.test.tsx]:
```suggestion
    const codeTextInput = getByTestId("code-text-input") as HTMLInputElement;
```

File: src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx (hunk 4 above)
Comment (mcmcphillips) [reply src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]: Why is an alias necessary here?

Comment (RohitKandimalla) [reply src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]: Yeah I did debugged about this a while I don't remember, but our jest setup is having issues identifying that module. 
I have seen that developers who use IntelliJ are seeing these warnings/errors, but devs who use VSCode doesn't get these suggestions, I haven't use VSCode so not sure what plugins we need to reflect our configs.

Comment (RohitKandimalla) [reply src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]: Yeah I remembered we need esModuleInterop flag to be set to true if we want to use default imports.

Comment (RohitKandimalla) [reply src/cqlEditorWithTerminology/CqlEditorWithTerminology.test.tsx]: image

File: src/terminologySectionPanel/codesSection/useCodeSystems.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  CodeSystem,
+} from "../../api/useTerminologyServiceApi";
+
+export function useCodeSystems() {
+  const [codeSystems, setCodeSystems] = useState<CodeSystem[]>([]);
+  // const [loading, setLoading] = useState(true); maybe add later since this call is slow
+
+  useEffect(() => {
+    const fetchCodeSystems = async () => {
+      // eslint-disable-next-line
```
Comment (sb-prateekkeerthi) [inline src/terminologySectionPanel/codesSection/useCodeSystems.ts:12]: we can remove these comments right ?

**Review by sb-prateekkeerthi — DISMISSED** [review]

File: src/terminologySectionPanel/codesSection/useCodeSystems.ts (hunk 14 above)
Comment (mcmcphillips) [reply src/terminologySectionPanel/codesSection/useCodeSystems.ts:12]: No, it breaks the build

Comment (mcmcphillips) [reply src/terminologySectionPanel/codesSection/useCodeSystems.ts:12]: the api class isn't actually an api class. It's a promise. 
The issue being thrown is a linter issue based off of naming conventions and the way we structure the api class. 

I tried a few variations of renaming files, functions, etc. But it appears to be completely just based on naming conventions. 
There didn't seem to be a way to fix this without completely rewriting the terminology api class to no longer be a promise.

File: src/api/useTerminologyServiceApi.ts [hunk 15]
```diff
@@ -110,6 +120,22 @@ export class TerminologyServiceApi {
...
+        `${this.baseUrl}/terminology/get-code-systems`,
+        {
+          headers: {
+            Authorization: `Bearer ${this.getAccessToken()}`,
+          },
+        }
+      );
+      return response.data;
+    } catch (err) {
+      console.error("Error retrieving getAllCodeSystems: ", err);
```
Comment (RohitKandimalla) [inline src/api/useTerminologyServiceApi.ts:136]: I think we need error handling here, Looks like if the getAllCodeSystems call fails, we just return an empty list. I wonder if we should let the user known that our call failed. But our story doesn't specify a negative scenario.

**Review by RohitKandimalla — APPROVED** [review]

File: src/api/useTerminologyServiceApi.ts (hunk 15 above)
Comment (RohitKandimalla) [reply src/api/useTerminologyServiceApi.ts:136]: You can resolve this, I just thought we may need some design decisions or questions to POs regarding the error handling.

**Pull request merged**